test/formula.jl

```diff
@@ -365,4 +368,38 @@ module TestFormula
     mm = ModelMatrix(mf)
     @test mm.m[:, 2] == d[complete_cases(d), :x1m]
 
+## Promote non-redundant categorical terms to full rank
+
+d = DataFrame(x = rep([:a, :b], times = 4),
+              y = rep([:c, :d], times = 2, each = 2),
+              z = rep([:e, :f], each = 4))
+[pool!(d, name) for name in names(d)]
+cs = [name => SumContrasts for name in names(d)]
+d[:n] = 1.:8
+
+
+## No intercept
+mf = ModelFrame(n ~ 0 + x, d[1:2,:], contrasts=cs)
```

Comment:
I think it would be clearer to use all rows, even it this indeed requires more typing. Also remove rows above 4, and `[1:4, :]`.

Testing complex cases would indeed be a good idea while you're at it. I'm afraid we won't return soon to this part of the code, in particular to check corner cases...